src/uct/api/uct.h

```diff
@@ -2065,17 +2185,12 @@ ucs_status_t uct_md_query(uct_md_h md, uct_md_attr_t *md_attr);
  * must support @ref UCT_MD_FLAG_ALLOC flag.
  *
  * @param [in]     md          Memory domain to allocate memory on.
- * @param [in,out] length_p    Points to the size of memory to allocate. Upon successful
- *                             return, filled with the actual size that was allocated,
- *                             which may be larger than the one requested. Must be >0.
- * @param [in,out] address_p   The address
- * @param [in]     flags       Memory allocation flags, see @ref uct_md_mem_flags.
- * @param [in]     name        Name of the allocated region, used to track memory
- *                             usage for debugging and profiling.
+ * @param [in]     param       Memory allocation characteristics, see
+ *                             @ref uct_mem_alloc_param_t.
  * @param [out]    memh_p      Filled with handle for allocated region.
  */
-ucs_status_t uct_md_mem_alloc(uct_md_h md, size_t *length_p, void **address_p,
-                              unsigned flags, const char *name, uct_mem_h *memh_p);
+ucs_status_t uct_md_mem_alloc(uct_md_h md, uct_mem_alloc_param_t *param,
```

Comment:
const uct_mem_alloc_param_t *param

Reply:
@yosefe Just realized that we can't use const qualifier because param->address_p is allowed to be changed after initially used a hint. Following two examples implement md_mem_alloc and return potentially new address_p values. For this reason we may have to change the param type to [inout] and remove const qualifier.

https://github.com/openucx/ucx/blob/master/src/uct/sm/mm/sysv/mm_sysv.c#L106
https://github.com/openucx/ucx/blob/master/src/uct/sm/mm/posix/mm_posix.c#L509

Reply:
not really, address_p is `void**`, not `void*`, we should be able to use const

Reply:
Oh right. I agree now that const can be used. 

What about setting param type from [in] to [inout] ? Don't we have to make this change given that param->address_p content may be updated by uct_md_mem_alloc?

Reply:
yeah, i though about it as well, but currently all our params (in both UCP and UCT) are "in" only, i think it's better to keep this convention

Reply:
Agreed. Then should we separate address pointed as hint and the returned address? So change the prototype to:
```
ucs_status_t uct_md_mem_alloc(uct_md_h md, uct_mem_alloc_param_t *param, void **address_p, uct_mem_h *memh_p);
```
and for uct_mem_alloc_params_t should the description change from:
```
    /**
     * If @a address_p content is NULL, the underlying allocation routine will
     * choose the address at which to create the mapping. If @a address_p
     * content is non-NULL but UCT_MD_MEM_FLAG_FIXED is not set, the address
     * will be interpreted as a hint as to where to establish the mapping. If
     * @a address_p is non-NULL and UCT_MD_MEM_FLAG_FIXED is set, then the
     * specified address is interpreted as a requirement. In this case, if the
     * mapping to the exact address cannot be made, the allocation request
     * fails.
     */
    void               **address_p;
```
to
```
    /**
     * If @a address content is NULL, the underlying allocation routine will
     * choose the address at which to create the mapping. If @a address
     * content is non-NULL but UCT_MD_MEM_FLAG_FIXED is not set, the address
     * will be interpreted as a hint as to where to establish the mapping. If
     * @a address is non-NULL and UCT_MD_MEM_FLAG_FIXED is set, then the
     * specified address is interpreted as a requirement. In this case, if the
     * mapping to the exact address cannot be made, the allocation request
     * fails.
     */
    void               *address;
```

Reply:
seems excessive to me.

Reply:
I'll leave it as is for now in that case.